### ISSUE TITLE
Chain functions is currently not supported

### ISSUE THREAD

**Issue by sagIoTPower** [IBM/JSONata4Java#211]
[Chaining of functions](https://docs.jsonata.org/programming#function-chaining) is currently not supported.

When I use the expression:
`$map(Customer, function($v, $i, $a) { $v ~> $uppercase ~> $trim } )`

for the json:

```
{
  "Customer": [
    "    paul   ",
    "  berverly     "
  ]
}
```

I run into an exception:
```
Exception in thread "main" java.lang.ClassCastException: class com.api.jsonata4java.expressions.generated.MappingExpressionParser$Var_recallContext 
cannot be cast to class com.api.jsonata4java.expressions.generated.MappingExpressionParser$Function_callContext (com.api.jsonata4java.expressions.generated.MappingExpressionParser$Var_recallContext 
and com.api.jsonata4java.expressions.generated.MappingExpressionParser$Function_callContext 
are in unnamed module of loader 'app')
        at com.api.jsonata4java.expressions.ExpressionsVisitor.visitFct_chain(ExpressionsVisitor.java:1654)
        at com.api.jsonata4java.expressions.ExpressionsVisitor.visitFct_chain(ExpressionsVisitor.java:99)
        at com.api.jsonata4java.expressions.generated.MappingExpressionParser$Fct_chainContext.accept(MappingExpressionParser.java:699)
```

Is it planned to implement this?

**Comment by wnm3** [comment]
There are no immediate plans to implement function chaining.

**Comment by bluemel** [comment]
Hmmmm... the chain operator seems to be kind of "half implemented".

1) it's already forseen in the ANTLR grammar
   - Line 134: CHAIN : '~>' ;
   - Line 70:   | expr CHAIN expr  # fct_chain

2) and we have also a vistor method for it in the ExpressionsVisitor class:
  - Line 1641: public JsonNode visitFct_chain(Fct_chainContext ctx) {

But it obviously does not work,
Hopefully I can find some time for debugging in the next days or weeks.

**Comment by bluemel** [comment]
On a new branch "#211-fct-chain" I've added some first JUnit tests with class *FunctionChainingTests*.
I started with testing with 2 functions "$sort()" and "$distinct()".
Suprisingly the chain already works out with the $sort() function but not with $dinstinct() (and probably also not with many other funcitions).
Nat, do you have a clue what the sort function does better than others?

**Comment by bluemel** [comment]
I seem to have fixed the chaining of the $distinct() function so far on branch #211-fct-chain but I fear that there are lot of functions left to bee checked... (e. g. $uppercase() or $trim())

**Comment by bluemel** [comment]
$distinct() and $sort() work out fine so far on my branch.

**Comment by wnm3** [comment]
That's great, I appreciate your working to fix this problem. 

Is there a basic issue in the functions that broke the chaining? 

There are function signatures that define the parameters for a function and whether they can work using the current context. So if a function can not read from context (e.g., only operates on its passed parameters) then they wouldn't work for chaining. These are few if any at this point.

**Comment by bluemel** [comment]
I can not yet speak for "functions" that broke the chaining because I've investigated so far only one of them which is ***distinct()**. So far as I understand the implementation of distinct has already foreseen context evaluation but simply did it wrong and crashed but was so far never (regression) tested if it works in a chain at all.

Now since I've simply changed the order of statements a bit according to how the **sort()** function implementation does it, it works out.

Please have a look at my changes on DistinctFunction.java.
https://github.com/IBM/JSONata4Java/compare/master...%23211-fct-chain

Probably I could have fixed it with less changes. I think the important changes concerning context evaluation are between line 60 and 80.

So concerning all other functions I fear we have to go through them one by one and see
* if they are intended to work in a chain
* if they work out properly

**Comment by wnm3** [comment]
I've tried looking but the changes are spread across multiple updates so it is difficult to follow. I believe one significant change was to support any kind of element, but I'm not sure what was actually causing the problem with chaining. You'd mentioned there was a problem using the context as input so perhaps the changes to count arguments and pull the context in as the first argument makes it "chainable?"

**Comment by bluemel** [comment]
Hi Nat, you can avoid the problem of vieving changes spread across multiple commits in github by pressing the "Compare and pull request" button of my branch (without pressing the "Create pull request" button). Then you have the acummulated view of all changes together.  

I've now reduced the changes on DistinctFunction.java to the minimal extent, so that function chaining also works with $distinct() (see  https://github.com/IBM/JSONata4Java/compare/%23211-fct-chain?expand=1)  

You see that I postponed the (premature) evaluation of ctx.exprValues().valuesCtx.exprList() done in lines 65-66 before to line 75 - to the case where argCount is 1 and useContext is false. I'd did this this way not because I understand really completely what's happening but because the implementation of $sort() does it this way. I'm not completely sure but relatively optimistic that we can leave the change as I made it - at least all yet existing tests are green. Maybe you have a better explanation for me why this works out.

**Comment by bluemel** [comment]
Concerning the original probelem reported in this issue I found out that his expression works out also in JSONata4Java when we add empty parantheses () to the function names:  

$map(Customer, function($v, $i, $a) { $v ~> $uppercase() ~> $trim() } ) 

So @sagIoTPower: you at least have a fine workround that will work out equally in JSONata4Java as well as in original JSONata.

**Comment by bluemel** [comment]
So let's state the following:
* JSONata4Java basically **does** support function chaining
* JSONata4Java does not yet support the "function name only" syntax for function chaining (the one without paratheses).
   Well... at least we could view this as a minor issue compared to not support the chaining at all.

**Comment by bluemel** [comment]
So how we should proceed with this issue?
We could either rename it to "Chaining of functions is currently not supported without parantheses" or similar or close it and open a new one with an appropriate title. Maybe the 2nd approach is even better because our discussion the could right from the start focus on the real problem.

**Comment by bluemel** [comment]
Concerning my branch that does not fix the problem reported here but fixes the chaining for $distinct() I'd propose to merge it and therefore opened a pull request for you, Nat. Maybe we could soon release a new version 2.2.2 containing this fix on $distinct.

**Comment by sagIoTPower** [comment]
> Concerning the original probelem reported in this issue I found out that his expression works out also in JSONata4Java when we add empty parantheses () to the function names:
> 
> $map(Customer, function($v, $i, $a) { $v ~> $uppercase() ~> $trim() } )
> 
> So @sagIoTPower: you at least have a fine workround that will work out equally in JSONata4Java as well as in original JSONata.

@bluemel 👍 Thank you very much for this advice. It already helps.

**Comment by wnm3** [comment]
Thank you -- merged the branch and will prepare to release 2.2.2 ASAP (sometime today).
